presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -837,6 +845,93 @@ public PlanNode visitSemiJoin(SemiJoinNode node, RewriteContext<Expression> cont
             return output;
         }
 
+        private PlanNode visitFilteringSemiJoin(SemiJoinNode node, RewriteContext<Expression> context)
+        {
+            Expression inheritedPredicate = context.get();
+            Expression deterministicInheritedPredicate = filterDeterministicConjuncts(inheritedPredicate);
+            Expression sourceEffectivePredicate = filterDeterministicConjuncts(EffectivePredicateExtractor.extract(node.getSource()));
+            Expression filteringSourceEffectivePredicate = filterDeterministicConjuncts(EffectivePredicateExtractor.extract(node.getFilteringSource()));
+            Expression joinExpression = new ComparisonExpression(
+                    ComparisonExpressionType.EQUAL,
+                    node.getSourceJoinSymbol().toSymbolReference(),
+                    node.getFilteringSourceJoinSymbol().toSymbolReference());
+
+            List<Symbol> sourceSymbols = node.getSource().getOutputSymbols();
+            List<Symbol> filteringSourceSymbols = node.getFilteringSource().getOutputSymbols();
+
+            List<Expression> sourceConjuncts = new ArrayList<>();
+            List<Expression> filteringSourceConjuncts = new ArrayList<>();
+            List<Expression> postJoinConjuncts = new ArrayList<>();
+
+            // Generate equality inferences
+            EqualityInference allInference = createEqualityInference(deterministicInheritedPredicate, sourceEffectivePredicate, filteringSourceEffectivePredicate, joinExpression);
+            EqualityInference allInferenceWithoutSourceInferred = createEqualityInference(deterministicInheritedPredicate, filteringSourceEffectivePredicate, joinExpression);
+            EqualityInference allInferenceWithoutFilteringSourceInferred = createEqualityInference(deterministicInheritedPredicate, sourceEffectivePredicate, joinExpression);
+
+            // Push inheritedPredicates down to the source if they don't involve the semi join output
+            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(inheritedPredicate)) {
+                Expression rewrittenConjunct = allInference.rewriteExpressionAllowNonDeterministic(conjunct, in(sourceSymbols));
+                // Since each source row is reflected exactly once in the output, ok to push non-deterministic predicates down
+                if (rewrittenConjunct != null) {
+                    sourceConjuncts.add(rewrittenConjunct);
+                }
+                else {
+                    postJoinConjuncts.add(conjunct);
+                }
+            }
+
+            // Push inheritedPredicates down to the filtering source if possible
+            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(deterministicInheritedPredicate)) {
+                Expression rewrittenConjunct = allInference.rewriteExpression(conjunct, in(filteringSourceSymbols));
+                // TODO can we push non-deterministic conjuncts there (not doing it for now)
```

Comment:
`can we ..` -- if this is supposed to be question, put `?` or `¿` somewhere

Reply:
most likely we can't

Reply:
So if we aren't sure about this, lets not leave this fuzzy TODO in the code to not to confuse people. Lets create an issue about that instead.

Reply:
No longer there https://github.com/prestodb/presto/commit/a85a9768b450a481e65a5c3bb7b3ab551719498b

Reply:
No longer there https://github.com/prestodb/presto/commit/a85a9768b450a481e65a5c3bb7b3ab551719498b